### ISSUE TITLE
Train on custom data

### ISSUE THREAD

**Issue by LeDuySon** [ifzhang/FairMOT#407]
I want to train on my custom dataset and i have some questions about that.
- What is the root path in cfg file ?
- What is the path for opt.data_dir 
- In my videos, some frames dont have any objects so when i use gen_labels_16.py, it generates only txt file for frame that have object, so i add some empty texts file for frames that dont have objects. Is my solution right? If not, how to deal with this.
My log when running with empty files:

![Screenshot from 2021-06-30 20-08-06](https://user-images.githubusercontent.com/33374938/123966558-bdca0580-d9df-11eb-86de-1cae5a7513ed.png)

**Comment by zengjie617789** [comment]
Apparently， the file path  "FairMOT/src/../dataset" is invalid, you should modify it.

**Comment by LeDuySon** [comment]
@zengjie617789 i think its not the problem because it can read my txt file but my txt file is empty so that the warning show up. Could u tell me how u deal with frames that dont have any object? Thanks!!!

**Comment by zengjie617789** [comment]
just to delete the ".." character in "src/../dataset" path, becauese of the path is invalid so it generate an  empty txt file.

**Comment by LeDuySon** [comment]
![Screenshot from 2021-07-01 10-59-23](https://user-images.githubusercontent.com/33374938/124062748-7d0fd200-da5b-11eb-9e70-e7df514deff5.png)

I have tried it but its not work! There is some emtpy files because  i add it for frames that dont have any object

**Comment by zengjie617789** [comment]
@LeDuySon  If there are some empty file, why not just remove them?   
I think your generaed labels are wrong. What's your content in label txt , post one for us  to check it.

**Comment by LeDuySon** [comment]
@zengjie617789  Some frames dont have any objects so its txt file is empty. If i dont add empty txt file for those frames, it will cause an error: Cannot find txt find with respect to frame...

Im using gen_labels_16.py to generate my txt file from gt.txt. Here is my folder:
![Screenshot from 2021-07-01 12-09-09](https://user-images.githubusercontent.com/33374938/124068892-87cf6480-da65-11eb-8285-c60fcd544475.png)

Content in label file:

![Screenshot from 2021-07-01 12-10-52](https://user-images.githubusercontent.com/33374938/124068930-93bb2680-da65-11eb-8653-4e530eb54311.png)

**Comment by zengjie617789** [comment]
Can i  understand you have many images and some image are no object in it, so their ground truth txt is empty？As a result, these empty file caused the error above? and you cannot simply remove these file, because these frame is exist?
My soulution to this is you should judge the gt conten when you generate the result.
my gt.txt is like blow:
```
585,75,758,481,63,29,0,9,0.48438
586,75,758,481,63,29,0,9,0.48438
587,75,758,481,63,29,0,9,0.48438
588,75,758,481,63,29,0,9,0.46875
```
as you described, you gt.txt may looks like:
```
586,75,758,481,63,29,0,9,0.48438
587
588,75,758,481,63,29,0,9,0.46875
```
 if so, you just add a judgement in here:

```
 for fid, tid, x, y, w, h, mark, label, _ in gt:
        if x or y or w or h is None:
            continue
        if mark == 0 or not label == 1:
            continue
```

**Comment by LeDuySon** [comment]
My gt.txt just like yours. And I think u misunderstand a little bit. Let me make it more clearly, when im using gen_labels_16.py to generate label files, frames that dont have any objects will not have txt file. So after that, i tried to train model but an error appeared because it cannot find the txt files for frames with 0 objects. To fix that i add some empty files for those frames. So i want to ask that my solution is right or not and if not what should i do? Should i drop frames that dont have object? And when we dont drop those frames, id loss will be NaN. Any solution?

Thanks for your help @zengjie617789 !

**Comment by zengjie617789** [comment]
I got it.  It seems that you have 588 images, but in the gt file  the line 587 is missing because of it detectes no one.
I think you absolutely should drop these images which detect none,  but it may be a heavy work to find the correspongding.
My suggestion is to regenerate the dataset. Just save  images people detected in.

**Comment by LeDuySon** [comment]
These images are frame in a video, its continuous so i think when we drop any frame in that, the tracks will not be updated and so do kalman? And maybe it will influence the final results.

**Comment by zengjie617789** [comment]
@LeDuySon ,  we will not using the kalman filter when we are training. You can see read_me.md file, the model is trained on the crowdhuman dataset which are combined of uncontinuous images.

**Comment by LeDuySon** [comment]
@zengjie617789 So it means we can remove frames that dont have object without affecting model performance. I try removing frames as you said and now i can train the model without any errors. Thank you so much!

I have 1 more question that how to finetune the model. My dataset is really small so i try freezing the backbone and train only heads but the results is pretty bad. The overall loss is 8.9 after 20 epochs with lr 1e-5. MOTA and IDF1 decrease 10% after finetuning. Do you have any suggestion?

**Comment by zengjie617789** [comment]
@LeDuySon ，emmmm， sorry, I think you should open another issue about this problem with offering more details about the dataset and params.
